sys/dev/sound/pcm/dsp.c

```diff
@@ -1502,24 +1502,30 @@ dsp_ioctl(struct cdev *i_dev, u_long cmd, caddr_t arg, int mode,
 
     	case SOUND_PCM_WRITE_CHANNELS:
 /*	case SNDCTL_DSP_CHANNELS: ( == SOUND_PCM_WRITE_CHANNELS) */
-		if (*arg_i < 0) {
+		if (*arg_i < 0 || *arg_i > AFMT_CHANNEL_MAX) {
 			*arg_i = 0;
 			ret = EINVAL;
 			break;
 		}
 		if (*arg_i != 0) {
-			struct pcmchan_matrix *m;
-			uint32_t ext;
+			uint32_t ext = 0;
 
 			tmp = 0;
-			if (*arg_i > SND_CHN_MAX)
-				*arg_i = SND_CHN_MAX;
+			/*
+			 * Map channel number to surround sound formats.
+			 * Devices with more than 8 channels need bitperfect
```

Comment:
more than `SND_CHN_MAX` channels?

(I've been wondering what would be required to increase this to support USB interfaces with > 8 channels.)

Reply:
@emaste, if I misread your comment, please elaborate - I hope it's clear why I'd want to avoid the limit of `SND_CHN_MAX` (== 8) in the bitperfect case?

Also, if you happen to know any committer interested in the sound infrastructure, please tell me. HPS' contributions are dearly missed in that department.

Reply:
I think this change is correct, I just mean that the comment should probably refer to `SND_CHN_MAX` (or e.g. `Devices with more than SND_CHN_MAX (8) channels`). If hypothetically we increased `SND_CHN_MAX` to 16 in the future it would invalidate the comment as written.

> This limit is needed for some of the feeders, matrix, equalizer and volume AFAICT.

Yes - perhaps increasing it beyond 8 isn't practical, but I'm sure it's not impossible. It may be that what's needed here is to have some tooling / opinionated setup for virtual_oss, so that there's a straightforward and standard approach. In any case, I'm just musing here, not arguing against this proposed patch.

There are a couple of developers who might take an interest in the sound infrastructure, but nothing concrete yet. In the interim I will try to review and shepherd in this change.

Reply:
Ok, got it, thanks for the clarification - and your efforts in general.

The updated comment should describe the situation more accurately now.